### PR TITLE
Kubernetes objects Sync status fix

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
Bug: https://bugzilla.redhat.com/show_bug.cgi?id=2274728

Few other spec changes:
1) protectedNamespace is renamed as protectedNamespaces. For more [Link](https://github.com/RamenDR/ramen/pull/1342)
2) Discovered apps namespace is renamed as openshift-dr-ops.  For more [Link](https://gitlab.cee.redhat.com/ceph/rhodf/-/merge_requests/2061/diffs#b696c0f64bdcba47d3a19c37f3b6ed9e1d06e8e1_62_62)

This PR need to be merged once after backend PR is merged: https://github.com/red-hat-storage/ramen/pull/244

### CODE REVIEW

**Comment by GowthamShanmugam** [comment]
image

File: locales/en/plugin__odf-console.json [hunk 1]
```diff
@@ -1301,12 +1300,28 @@
...
+  "exists": "exists",
+  "does not exist": "does not exist",
+  "Expand to enter expression": "Expand to enter expression",
+  "Label": "Label",
+  "Select a label": "Select a label",
+  "Operator": "Operator",
+  "Values": "Values",
+  "Select the values": "Select the values",
+  "Add label expression": "Add label expression",
   "Delete {{kind}}?": "Delete {{kind}}?",
```
Comment (TimothyAsirJeyasing) [inline locales/en/plugin__odf-console.json:1325]: Are we using these in this PR

File: packages/mco/components/protected-applications/components.tsx [hunk 2]
```diff
@@ -382,9 +383,16 @@ export const StatusDetails: React.FC<ExpandableComponentProps> = ({
...
+    <Text className="text-muted"> {t('No data available')}</Text>
+  );
+
   const lastSyncOn = [
-    syncStatusInfo.volumeLastGroupSyncTime,
-    syncStatusInfo.kubeObjectLastSyncTime,
+    syncStatusInfo.volumeLastGroupSyncTime || noDataMessage,
+    // For kube object, it is always no data available message.
+    // lastTransitionTime just to decided the status, it can't be considered as last synced on.
+    noDataMessage,
```
Comment (TimothyAsirJeyasing) [inline packages/mco/components/protected-applications/components.tsx:395]: It used one time only, do we need a separate variable ?

File: packages/mco/utils/disaster-recovery.tsx [hunk 3]
```diff
@@ -636,3 +646,20 @@ export const getLabelsFromSearchResult = (
...
+): string => {
+  const condition = getConditionFromStatus(
+    conditions,
+    KUBE_OBJECT_SYNC_CONDITION_TYPE
+  );
+  if (condition?.status === 'True') {
+    // For "True", checking the lastTransitionTime to ensure there is no reconcile problem on ramen.
+    // If there is any reconcile problem, And if status stuck in "True"
+    // then status will be displayed based on lastTransitionTime (critical/warning).
+    return condition?.lastTransitionTime;
```
Comment (TimothyAsirJeyasing) [inline packages/mco/utils/disaster-recovery.tsx:661]: How about returning empty string if lastTransitionTime is undefined
return condition.lastTransitionTime || '';

**Review by TimothyAsirJeyasing — COMMENTED** [review]
LGTM with some minor comments

File: locales/en/plugin__odf-console.json (hunk 1 above)
Comment (GowthamShanmugam) [reply locales/en/plugin__odf-console.json:1325]: its a line change in translation file: https://github.com/red-hat-storage/odf-console/pull/1295/files#diff-231cc3a06bdedaa5b04414ae58071e28afb38c02dfd29379a3ea03febd69acf0L1403

File: packages/mco/components/protected-applications/components.tsx (hunk 2 above)
Comment (GowthamShanmugam) [reply packages/mco/components/protected-applications/components.tsx:395]: it is used twice lin no 392 and 395

File: packages/mco/utils/disaster-recovery.tsx (hunk 3 above)
Comment (GowthamShanmugam) [reply packages/mco/utils/disaster-recovery.tsx:661]: condition?.status === 'True' means mostly lastTransitionTime is non empty, may be I could remove the optional symbol

File: packages/mco/constants/disaster-recovery.ts [hunk 4]
```diff
@@ -98,3 +98,6 @@ export const EnrollDiscoveredApplicationStepNames = (t: TFunction) => ({
   [EnrollDiscoveredApplicationSteps.Replication]: t('Replication'),
   [EnrollDiscoveredApplicationSteps.Review]: t('Review'),
 });
+
+// DRPC status condition types
+export const KUBE_OBJECT_SYNC_CONDITION_TYPE = 'ClusterDataProtected';
```
Comment (SanjalKatiyar) [inline packages/mco/constants/disaster-recovery.ts:103]: sounds like generic name... is this not specifically related to KubeObjects ??

File: packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx [hunk 5]
```diff
@@ -121,10 +122,10 @@ export const useDiscoveredParser: UseDiscoveredParser = (
                   kubeObjSyncInterval:
                     drPlacementControl.spec?.kubeObjectProtection
                       ?.captureInterval,
-                  lastKubeObjectSyncTime:
-                    // ToDo: Update with correct status field which will report kube object last sync time
-                    // @ts-ignore
-                    drPlacementControl?.status?.lastKubeObjectSyncTime,
+                  // lastTransitionTime just to decided the status, it can't be considered as last synced on.
```
Comment (SanjalKatiyar) [inline packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx:125]:
```suggestion
                  // lastTransitionTime just to decide the status, it can't be considered as last synced on.
```

File: packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx [hunk 6]
```diff
@@ -121,10 +122,10 @@ export const useDiscoveredParser: UseDiscoveredParser = (
...
+                  lastKubeObjectSyncTime: getKubeObjectLastSyncTime(
```
Comment (SanjalKatiyar) [inline packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx:126]: if that's the case we might want to rename here as well ??
```suggestion
                  kubeObjectLastTransitionTime: getKubeObjectLastTransitionTime(
```

File: packages/mco/components/protected-applications/components.tsx [hunk 7]
```diff
@@ -382,9 +383,16 @@ export const StatusDetails: React.FC<ExpandableComponentProps> = ({
...
+  const noDataMessage = (
+    <Text className="text-muted"> {t('No data available')}</Text>
+  );
+
   const lastSyncOn = [
-    syncStatusInfo.volumeLastGroupSyncTime,
-    syncStatusInfo.kubeObjectLastSyncTime,
+    syncStatusInfo.volumeLastGroupSyncTime || noDataMessage,
+    // For kube object, it is always no data available message.
+    // lastTransitionTime just to decided the status, it can't be considered as last synced on.
```
Comment (SanjalKatiyar) [inline packages/mco/components/protected-applications/components.tsx:394]:
```suggestion
    // lastTransitionTime just to decide the status, it can't be considered as last synced on.
```

File: packages/mco/components/protected-applications/list-page.tsx [hunk 8]
```diff
@@ -260,20 +264,20 @@ export const ProtectedApplicationsListPage: React.FC = () => {
...
-        const kubeObjectsLastSyncTime = app?.status?.lastKubeObjectSyncTime;
+        const kubeObjectsLastSyncTime = getKubeObjectLastSyncTime(
+          app?.status?.resourceConditions?.conditions
+        );
         acc[getName(app)] = {
           volumeReplicationStatus: getReplicationHealth(
             volumesLastSyncTime,
             volumesSchedulingInterval
           ),
-          volumeLastGroupSyncTime: formatTime(volumesLastSyncTime) || DASH,
```
Comment (SanjalKatiyar) [inline packages/mco/components/protected-applications/list-page.tsx:271]: why removing `DASH` ??

File: packages/mco/utils/disaster-recovery.tsx [hunk 9]
```diff
@@ -551,14 +553,22 @@ export const findPlacementDecisionUsingPlacement = (
       getNamespace(placementDecision) === getNamespace(placement)
   );
 
+export const getConditionFromStatus = (
+  conditions: K8sResourceCondition[],
+  conditionType: string
+): K8sResourceCondition =>
+  conditions?.find((condition) => condition.type === conditionType);
+
```
Comment (SanjalKatiyar) [inline packages/mco/utils/disaster-recovery.tsx:561]: we have `getResourceCondition` under `k8s.ts`...

File: packages/mco/components/protected-applications/list-page.tsx (hunk 8 above)
Comment (GowthamShanmugam) [reply packages/mco/components/protected-applications/list-page.tsx:271]: As per UXchange, I want "No data available" text instead of a dash.

File: packages/mco/constants/disaster-recovery.ts (hunk 4 above)
Comment (GowthamShanmugam) [reply packages/mco/constants/disaster-recovery.ts:103]: Its for all metadata backup + kubeObject backend, for now we dont have any dedicated status for kube object.

File: packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx (hunk 6 above)
Comment (GowthamShanmugam) [reply packages/mco/components/mco-dashboard/disaster-recovery/parsers/discovered-parser.tsx:126]: makesense

File: packages/mco/utils/disaster-recovery.tsx (hunk 9 above)
Comment (GowthamShanmugam) [reply packages/mco/utils/disaster-recovery.tsx:561]: this is for DRPC specific drPlacementControl?.status?.resourceConditions?.conditions

File: packages/mco/utils/disaster-recovery.tsx [hunk 10]
```diff
@@ -636,3 +638,19 @@ export const getLabelsFromSearchResult = (
...
 };
+
+export const getKubeObjectLastTransitionTime = (
+  conditions: K8sResourceCondition[]
+): string => {
+  const condition = conditions?.find(
+    (condition) => condition.type === KUBE_OBJECT_SYNC_CONDITION_TYPE
+  );
+  if (condition?.status === 'True') {
+    // For "True", checking the lastTransitionTime to ensure there is no reconcile problem on ramen.
```
Comment (TimothyAsirJeyasing) [inline packages/mco/utils/disaster-recovery.tsx:649]: How about "no reconciliation issues with ramen" ..

Comment (GowthamShanmugam) [reply packages/mco/utils/disaster-recovery.tsx:649]: ya it's just a comment, it is good as long as other developers can understand why this change

Comment (TimothyAsirJeyasing) [reply packages/mco/utils/disaster-recovery.tsx:649]: Ok

**Comment by GowthamShanmugam** [comment]
/ hold cancel

File: packages/mco/components/protected-applications/components.tsx [hunk 11]
```diff
@@ -382,9 +383,16 @@ export const StatusDetails: React.FC<ExpandableComponentProps> = ({
       {kubeIcon} {kubeTitle}
     </>,
   ];
+
+  const noDataMessage = (
+    <Text className="text-muted"> {t('No data available')}</Text>
+  );
```
Comment (bipuladh) [inline packages/mco/components/protected-applications/components.tsx:389]: This is better suited as a React.FC component

**Comment by GowthamShanmugam** [comment]
Condition types differ for each CR, And each CR can be from different projects. We can't maintain every condition type in shared. I would say instead of keeping this function-specific, it should be generic.

I don't like declaring all of my DR-related condition types in a shared repo to use this function. It is better to maintain one more util function in the DR console repo.

File: packages/shared/src/selectors/k8s.ts [hunk 12]
```diff
@@ -62,7 +62,7 @@ export const getResourceCondition = <
   A extends K8sResourceKind = K8sResourceKind
 >(
   resource: A,
-  condition: KnownResourceConditions
+  condition: KnownResourceConditions | string
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply packages/shared/src/selectors/k8s.ts]: Condition types differ for each CR, And each CR can be from different projects. We can't maintain every condition type in shared. I would say instead of keeping this function-specific, it should be generic.

I don't like declaring all of my DR-related condition types in a shared repo to use this function. It is better to maintain one more util function in the DR console repo.
